zcash_client_backend/src/address.rs

```diff
@@ -1,18 +1,151 @@
+        constants::mainnet::HRP_SAPLING_PAYMENT_ADDRESS => Network::Main,
+        constants::testnet::HRP_SAPLING_PAYMENT_ADDRESS => Network::Test,
+        constants::regtest::HRP_SAPLING_PAYMENT_ADDRESS => Network::Regtest,
+        _ => panic!("Unsupported network kind"),
+    }
+}
+
+/// A Unified Address.
+#[derive(Clone, Debug, PartialEq)]
+pub struct UnifiedAddress {
+    orchard: Option<orchard::Address>,
+    sapling: Option<PaymentAddress>,
+    transparent: Option<TransparentAddress>,
+    unknown: Vec<(u32, Vec<u8>)>,
+}
+
+impl TryFrom<unified::Address> for UnifiedAddress {
+    type Error = &'static str;
+
+    fn try_from(ua: unified::Address) -> Result<Self, Self::Error> {
+        let mut orchard = None;
+        let mut sapling = None;
+        let mut transparent = None;
+
+        // We can use as-parsed order here for efficiency, because we're breaking out the
+        // receivers we support from the unknown receivers.
+        let unknown = ua
+            .items_as_parsed()
+            .iter()
+            .filter_map(|receiver| match receiver {
+                unified::Receiver::Orchard(data) => {
+                    Option::from(orchard::Address::from_raw_address_bytes(data))
+                        .ok_or("Invalid Orchard receiver in Unified Address")
+                        .map(|addr| {
+                            orchard = Some(addr);
+                            None
+                        })
+                        .transpose()
+                }
+                unified::Receiver::Sapling(data) => PaymentAddress::from_bytes(data)
+                    .ok_or("Invalid Sapling receiver in Unified Address")
+                    .map(|pa| {
+                        sapling = Some(pa);
+                        None
+                    })
+                    .transpose(),
+                unified::Receiver::P2pkh(data) => {
+                    transparent = Some(TransparentAddress::PublicKey(*data));
+                    None
+                }
+                unified::Receiver::P2sh(data) => {
+                    transparent = Some(TransparentAddress::Script(*data));
+                    None
+                }
+                unified::Receiver::Unknown { typecode, data } => {
+                    Some(Ok((*typecode, data.clone())))
+                }
+            })
+            .collect::<Result<_, _>>()?;
+
+        Ok(Self {
+            orchard,
+            sapling,
+            transparent,
+            unknown,
+        })
+    }
+}
+
+impl UnifiedAddress {
+    /// Constructs a Unified Address from a given set of receivers.
+    ///
+    /// Returns `None` if the receivers would produce an invalid Unified Address (namely,
+    /// if no shielded receiver is provided).
+    pub fn from_receivers(
+        orchard: Option<orchard::Address>,
+        sapling: Option<PaymentAddress>,
+        transparent: Option<TransparentAddress>,
+    ) -> Option<Self> {
```

Comment:
This could be conceptually wrong because I'm a Rust newbie. 

Should this be a wrapper a function that actually takes the unknown receiver as well?

````Rust
 pub fn from_receivers(
        orchard: Option<orchard::Address>,
        sapling: Option<PaymentAddress>,
        transparent: Option<TransparentAddress>,
        unknown: Vec<(u32, Vec<u8>)>,
    ) -> Option<Self> {
````

Reply:
"No-ish"

This is currently sort of a hybrid of two structs: a receiver-aware struct that can handle any address, and a receiver-aware struct that is only for wallet-generated addresses. I currently don't want to expose the ability to construct a new `UnifiedAddress` with arbitrary receivers (I only want known receivers), hence `unknown` is not exposed. However `unknown` is necessary in order to _parse_ arbitrary UAs, since we need to store them to ensure round-trip serialization.

This API will likely end up changing once I've dug us out from under this particular tech debt hill, but for now I want to keep things as minimal and straightforward as possible while we go through the multiple PRs to fix the structure of `UnifiedFullViewingKey`.